src/v/cluster/types.cc

```diff
@@ -298,11 +322,22 @@ void adl<cluster::topic_configuration>::to(
       t.properties.timestamp_type,
       t.properties.segment_size,
       t.properties.retention_bytes,
-      t.properties.retention_duration);
+      t.properties.retention_duration,
+      t.properties.recovery,
+      t.properties.shadow_indexing);
 }
 
 cluster::topic_configuration
 adl<cluster::topic_configuration>::from(iobuf_parser& in) {
+    auto version = adl<int32_t>{}.from(in.peek(4));
```

Comment:
A bit unpleasant that we have to allocate a new buffer just to parse an int. Maybe consume it and (in the case of old format) pass the length to the `ns` parser?

Reply:
I guess we could add `adl::peek` but that's a bit more work.